x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerRules.java

```diff
@@ -1137,6 +1141,49 @@ protected Expression rule(Expression e) {
         }
     }
 
+    public static final class SimplifyArithmeticsInBinaryComparisons extends OptimizerExpressionRule {
+
+        public SimplifyArithmeticsInBinaryComparisons() {
+            super(TransformDirection.UP);
+        }
+
+        @Override
+        protected Expression rule(Expression e) {
+            if (e instanceof BinaryComparison) {
+                return reduce((BinaryComparison) e);
+            }
+            return e;
+        }
+
+        private Expression reduce(BinaryComparison bc) {
+            // optimize only once the expression looks like: AnyNode [ArithmOp] AnyNode [BiComp] Literal
+            if (bc.right() instanceof Literal == false) {
+                return bc;
+            }
+            Literal bcRight = (Literal) bc.right();
+
+            if (bc.left() instanceof ArithmeticOperation) {
+                ArithmeticOperation opLeft = (ArithmeticOperation) bc.left();
+                if (opLeft instanceof Mod) {
+                    return bc; // can't optimise Mods
+                }
+
+                if (opLeft.left() instanceof Literal || opLeft.right() instanceof Literal) { // 5 [op] a >= 4 || a [op] 5 >= 4
+                    // force double division
+                    if (opLeft.symbol().equals("*")) { // use symbol comp for SQL's Mul
```

Comment:
Try to avoid using strings for type matching. Assuming there's no other either way move use a string reference to indicate its source, such as `DefaultBinaryArithmeticOperation.MUL.symbol()` instead of `*`.

Also please add some comments to explain what's going on - if the symbol is Mul, force double division by promoting the right literal to a double?

Reply:
>  such as DefaultBinaryArithmeticOperation.MUL.symbol() instead of *.

Thanks, fixed.

> if the symbol is Mul, force double division by promoting the right literal to a double?

Yes, that was the logic. But this got now removed, as a potentially unsafe optimisation.